versions/development/SPEC.md

```diff
@@ -1218,81 +1256,343 @@ $runtime = 'runtime' $ws* '{' ($ws* $runtime_kv $ws*)* '}'
+
+Container source locations should use the syntax defined by the individual container repository. For example an image defined as `ubuntu:latest` would conventionally refer a docker image living on `DockerHub`, while an image defined as `quay.io/bitnami/python` would refer to a `quay.io` repository. 
 
+The supported attribute types are:
+
+* `String`: A single container location
+* `Array[String]`: A list of container entries. **Note**: the ordering of the list does not imply any implicit preference or ordering of the containers. All images are expected to be the same, and therefore any choice would be equally valid.
+
+**Single Image**
 ```wdl
-task test {
-  input {
-    String ubuntu_version
+task single_image_test {
+  #....
+  runtime {
+    container: "ubuntu:latest"
   }
-  command {
-    python script.py
+```
+
+**Multiple Images**
+```wdl
+task multiple_image_test {
+  #.....
+  runtime {
+    container: ["ubuntu:latest", "gcr.io/standard-images/ubuntu:latest"]
   }
+}
+```
+
+
+#### cpu (**Optional**)
+
+The `cpu` key defines the _minimum_ CPU required for this task, which must be available prior to the engine starting execution. The engine does not need to provide the exact amount of CPU requested (depending on the underlying infrastructure restrictions), however it may ONLY provision more CPU than requested and not less. For example if the wdl requested `cpu: 0.5`, but only discrete values were supported, then the engine might choose to provision `1.0` cpu instead.  Values are expected to be a `Float` or an `Int`
+
+
+##### Default Value
+
+If the `cpu` key is not defined, then the engine must set a default value of `1` CPU. The task, therefore will be guaranteed to run with 1 CPU. The engine must not provide a mechanism to set their own default value for disks which circumvents this, ie `runtime-defaults`. 
+
+
+```wdl
+task cpu_example {
+  #....	
   runtime {
-    docker: "ubuntu:" + ubuntu_version
+    cpu: 8
   }
 }
+
 ```
 
-Most key/value pairs are arbitrary.  However, the following keys have recommended conventions:
+#### memory (**Optional**)
 
-#### docker
+The `memory` key defines the _minimum_ memory required for this task which must be available prior to the engine starting execution. The engine does not need to provide the exact amount of memory requested, however it may ONLY provision more memory then requested and not less. For example, if the wdl requested `1 GB` but only blocks of `4 GB` were available, the engine might choose to provision `4.0 GB` instead.  Two kinds of values are supported for this attribute:
```

Comment:
One side effect of a one-size-fits-all `memory` attribute which I've spotted is people end up having to write a "safe" amount of memory here, rather than the "should work 80% of the time" amount (and allow the engine to try again with more if it needs to).

My gut feeling is that the definition of memory here should actually be something more like "feel free to try with less memory at first if you really want to, but if you give the tool this much memory and it still fails, fail the task". My thinking is:
* The simplest engine implementation is to provide `memory` and fail if the task doesn't succeed
* More complicated engine implementations still get a ceiling that they know that it's not worth going above

Is that fair?

Reply:
One size fits all memory is an issue that I have come across to, anmd the only reliable solution I have found is generally trial and error. I think this is a difficult point. How do we define "less memory" . Trying with less and resulting in a failure could be a costly mistake that the user now has to pay for if they are in a cloud environment. 

To me resource requirements are minimum guarantees. As a task creator I 100% known that my Task will minimally be given the requested memory, and therefore I can design the task with that in mind. The idea of course is I have tested the command I am running in multiple ways (this might be alot to ask), and have Identified the appropriate amount of memory needed. The `maxMemory` is a flag which they can use in the `hints` section to specify the upper bound that it should be given. Allowing memory to be below this removes any sort of guarantees and now makes the execution of the task a bit unpredictable.

Reply:
If it task runs out of memory, I am not sure itsd the responsibility of the execution engine to retry that task. It might be, but you could also argue that it is up to the user whether or not they want that retry to take place

Reply:
@cjllanwarne With user defined overrides that same behavior is captured IMO